src/lib.rs

```diff
@@ -268,27 +385,35 @@ pub trait Arbitrary<'a>: Sized {
     ///     }
     ///
     ///     fn size_hint(depth: usize) -> (usize, Option<usize>) {
+    ///         // Return the value of faillible_size_hint
+    ///         //
+    ///         // If the recursion fails, return the default, always valid `(0, None)`
+    ///         Self::faillible_size_hint(depth).unwrap_or_default()
+    ///     }
+    ///
+    ///     fn faillible_size_hint(depth: usize) -> Result<(usize, Option<usize>), MaxRecursionReached> {
     ///         // Protect against potential infinite recursion with
-    ///         // `recursion_guard`.
-    ///         size_hint::recursion_guard(depth, |depth| {
-    ///             // If we aren't too deep, then `recursion_guard` calls
-    ///             // this closure, which implements the natural size hint.
-    ///             // Don't forget to use the new `depth` in all nested
-    ///             // `size_hint` calls! We recommend shadowing the
-    ///             // parameter, like what is done here, so that you can't
-    ///             // accidentally use the wrong depth.
-    ///             size_hint::or(
-    ///                 <L as Arbitrary>::size_hint(depth),
-    ///                 <R as Arbitrary>::size_hint(depth),
-    ///             )
+    ///         // `faillible_recursion_guard`.
+    ///         size_hint::faillible_recursion_guard(depth, |depth| {
+    ///            // If we aren't too deep, then `recursion_guard` calls
+    ///            // this closure, which implements the natural size hint.
+    ///            // Don't forget to use the new `depth` in all nested
+    ///            // `faillible_size_hint` calls! We recommend shadowing the
+    ///            // parameter, like what is done here, so that you can't
+    ///            // accidentally use the wrong depth.
+    ///            Ok(size_hint::or(
+    ///                <L as Arbitrary>::faillible_size_hint(depth)?,
+    ///                <R as Arbitrary>::faillible_size_hint(depth)?,
+    ///            ))
     ///         })
     ///     }
     /// }
     /// ```
+    /// [iterator-size-hint]: https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.size_hint
+    /// [`size_hint`]: Arbitrary::size_hint
     #[inline]
-    fn size_hint(depth: usize) -> (usize, Option<usize>) {
-        let _ = depth;
-        (0, None)
+    fn faillible_size_hint(depth: usize) -> Result<(usize, Option<usize>), MaxRecursionReached> {
```

Comment:
Nitpick: "faillible" is a French word, the English is "fallible". So we should rename this method to `fallible_size_hint` or perhaps `try_size_hint`.

Reply:
`try_size_hint` is better indeed.